File: product/frontend/src/api/projects.ts
/* =============================================================================
 * 应用与项目 API 客户端
 *
 * 定位
 *   接入页面与 Project HTTP 路由之间的前端能力适配器
 *
 * 职责
 *   建立应用连接｜确认 endpoint｜授权源码分析｜读取当前 ApplicationUnderstanding
 *
 * 调用链
 *   Access / ControlShell → projectsApi → api/http
 * ============================================================================= */

import { request } from './http'
import type { DiscoveryResult } from './onboarding'

export type ProjectDto = {
  project_id: string
  name?: string
  status?: string
  governed_contract_id?: string | null
  governed_contract_version?: number | null
  created_at_us?: number
  updated_at_us?: number
}

export type ApplicationUnderstandingDto = {
  project_id: string
  source_root: string
  confirmed_endpoint: string | null
  endpoint_source_fingerprint: string | null
  endpoint_confirmed_at_us: number | null
  endpoint_last_checked_at_us: number | null
  endpoint_reachable: boolean | null
  source_analysis_authorized: boolean
  source_analysis_authorized_at_us: number | null
  source_fingerprint: string | null
  analysis_completed_at_us: number | null
  role_candidates: RoleCandidateDto[]
  action_candidates: ActionCandidateDto[]
  revision: number
  created_at_us: number
  updated_at_us: number
}

export type CandidateEvidenceDto = {
  relative_path: string
  line_start: number
  line_end: number
  symbol: string | null
  detector: string
  content_sha256: string
}

export type RoleCandidateDto = {
  candidate_id: string
  canonical_key: string
  display_name: string
  confidence: 'HIGH' | 'MEDIUM' | 'LOW'
  decision: 'PROPOSED' | 'CONFIRMED' | 'REJECTED' | 'REVIEW_REQUIRED'
  origin: 'DETECTED' | 'MANUAL'
  stale: boolean
  evidence: CandidateEvidenceDto[]
}

export type ActionCandidateDto = RoleCandidateDto & {
  risk_hint: 'READ' | 'WRITE' | 'DELETE' | 'ADMIN' | 'UNKNOWN'
}

export type EndpointCandidateDto = {
  endpoint: string
  source_type: 'CONFIG' | 'OPENAPI' | 'STARTUP' | 'FRAMEWORK_DEFAULT'
  source: string
  rank: number
  reachable: boolean
  status_code: number | null
  probe_detail: string
  confirmation_required: true
}

export type EndpointDiscoveryDto = {
  source_fingerprint: string
  candidates: EndpointCandidateDto[]
  request_count: number
  default_endpoint: string | null
  manual_entry_required: boolean
}

export type ApplicationConnectionDto = {
  project: ProjectDto
  understanding: ApplicationUnderstandingDto
  discovery: DiscoveryResult
}

export const projectsApi = {
  projects: () => request<ProjectDto[]>('/api/projects'),
  archivedProjects: () => request<ProjectDto[]>('/api/projects?include_archived=true'),
  connectApplication: (sourceRoot: string, projectName?: string) =>
    request<ApplicationConnectionDto>('/api/applications/connect', {
      method: 'POST',
      body: JSON.stringify({ schema_version: '1', source_root: sourceRoot, project_name: projectName }),
    }),
  project: (id: string) => request<ProjectDto>(`/api/projects/${id}`),
  remove: (id: string) => request<ProjectDto>(`/api/projects/${id}`, { method: 'DELETE' }),
  understanding: (id: string) => request<ApplicationUnderstandingDto>(`/api/projects/${id}/application-understanding`),
  discoverEndpoints: (id: string) => request<EndpointDiscoveryDto>(`/api/projects/${id}/endpoint-candidates`, { method: 'POST' }),
  confirmEndpoint: (id: string, endpoint: string, revision: number) =>
    request<ApplicationUnderstandingDto>(`/api/projects/${id}/endpoint`, {
      method: 'PUT',
      body: JSON.stringify({ schema_version: '1', endpoint, revision }),
    }),
  authorizeSourceAnalysis: (id: string, revision: number) =>
    request<ApplicationUnderstandingDto>(`/api/projects/${id}/source-analysis-authorization`, {
      method: 'PUT',
      body: JSON.stringify({ schema_version: '1', authorized: true, revision }),
    }),
  analyzeSource: (id: string, revision: number) =>
    request<ApplicationUnderstandingDto>(`/api/projects/${id}/source-analysis`, {
      method: 'POST',
      body: JSON.stringify({ schema_version: '1', revision }),
    }),
  decideRole: (id: string, candidateId: string, decision: 'PROPOSED' | 'CONFIRMED' | 'REJECTED', displayName: string, revision: number) =>
    request<ApplicationUnderstandingDto>(`/api/projects/${id}/roles/${candidateId}`, {
      method: 'PUT',
      body: JSON.stringify({ schema_version: '1', decision, display_name: displayName, revision }),
    }),
  addRole: (id: string, displayName: string, revision: number) =>
    request<ApplicationUnderstandingDto>(`/api/projects/${id}/roles`, {
      method: 'POST',
      body: JSON.stringify({ schema_version: '1', display_name: displayName, revision }),
    }),
  decideAction: (id: string, candidateId: string, decision: 'PROPOSED' | 'CONFIRMED' | 'REJECTED', displayName: string, revision: number) =>
    request<ApplicationUnderstandingDto>(`/api/projects/${id}/actions/${candidateId}`, {
      method: 'PUT',
      body: JSON.stringify({ schema_version: '1', decision, display_name: displayName, revision }),
    }),
  addAction: (id: string, displayName: string, riskHint: ActionCandidateDto['risk_hint'], revision: number) =>
    request<ApplicationUnderstandingDto>(`/api/projects/${id}/actions`, {
      method: 'POST',
      body: JSON.stringify({ schema_version: '1', display_name: displayName, risk_hint: riskHint, revision }),
    }),
}

File: product/backend/api/routers/projects.py
# 项目 API 路由：暴露应用接入、理解候选与当前权限合同的受控 HTTP 边界。
# 适配 Project 接入和当前 PermissionContract 读取，不在路由层推断治理结论。

from __future__ import annotations

from typing import Literal
from fastapi import APIRouter
from pydantic import Field

from product.backend.workflows.context import ApplicationCore
from product.backend.core.application_understanding import (
    ActionRiskHint,
    CandidateDecision,
)
from product.backend.api.envelope import data_response
from product.backend.api.envelope import ApiResponse
from product.backend.api.envelope import ApiModel


def build_projects_router(context: ApplicationCore) -> APIRouter:
    router = APIRouter()

    @router.post("/api/applications/connect", response_model=ApiResponse)
    def connect_application(body: ApplicationConnectRequest):
        result = context.application_understanding.connect(
            body.source_root,
            project_name=body.project_name,
        )
        return data_response(result.model_dump(mode="json"), status_code=201)

    @router.get("/api/projects", response_model=ApiResponse)
    async def list_projects(include_archived: bool = False):
        return data_response(
            [
                record.model_dump(mode="json")
                for record in context.projects.list(include_archived=include_archived)
            ]
        )

    @router.get("/api/projects/{project_id}", response_model=ApiResponse)
    async def get_project(project_id: str):
        return data_response(context.projects.get(project_id).model_dump(mode="json"))

    @router.delete("/api/projects/{project_id}", response_model=ApiResponse)
    async def archive_project(project_id: str):
        """移除普通应用视图，保留 Project 及全部历史结果。"""

        return data_response(
            context.project_lifecycle.archive(project_id).model_dump(mode="json")
        )

    @router.get("/api/projects/{project_id}/status", response_model=ApiResponse)
    async def get_product_status(project_id: str):
        """返回 GUI、CLI 与 Machine 共用的项目工作台只读投影。"""

        return data_response(
            context.product_status.get(project_id).model_dump(mode="json")
        )

    @router.get(
        "/api/projects/{project_id}/application-understanding",
        response_model=ApiResponse,
    )
    def get_application_understanding(project_id: str):
        return data_response(
            context.application_understanding.get(project_id).model_dump(mode="json")
        )

    @router.post(
        "/api/projects/{project_id}/endpoint-candidates",
        response_model=ApiResponse,
    )
    def discover_endpoint_candidates(project_id: str):
        return data_response(
            context.application_understanding.discover_endpoints(project_id).model_dump(
                mode="json"
            )
        )

    @router.put(
        "/api/projects/{project_id}/endpoint",
        response_model=ApiResponse,
    )
    def confirm_endpoint(project_id: str, body: EndpointConfirmationRequest):
        return data_response(
            context.application_understanding.confirm_endpoint(
                project_id,
                endpoint=body.endpoint,
                revision=body.revision,
            ).model_dump(mode="json")
        )

    @router.put(
        "/api/projects/{project_id}/source-analysis-authorization",
        response_model=ApiResponse,
    )
    def authorize_source_analysis(
        project_id: str,
        body: SourceAnalysisAuthorizationRequest,
    ):
        return data_response(
            context.application_understanding.authorize_source_analysis(
                project_id,
                revision=body.revision,
            ).model_dump(mode="json")
        )

    @router.post(
        "/api/projects/{project_id}/source-analysis",
        response_model=ApiResponse,
    )
    def analyze_source(project_id: str, body: SourceAnalysisRequest):
        return data_response(
            context.application_understanding.analyze_source(
                project_id,
                revision=body.revision,
            ).model_dump(mode="json")
        )

    @router.put(
        "/api/projects/{project_id}/roles/{candidate_id}",
        response_model=ApiResponse,
    )
    def decide_role(
        project_id: str,
        candidate_id: str,
        body: CandidateDecisionRequest,
    ):
        return data_response(
            context.application_understanding.decide_role(
                project_id,
                candidate_id,
                revision=body.revision,
                decision=CandidateDecision(body.decision),
                display_name=body.display_name,
            ).model_dump(mode="json")
        )

    @router.post(
        "/api/projects/{project_id}/roles",
        response_model=ApiResponse,
    )
    def add_manual_role(project_id: str, body: ManualRoleRequest):
        return data_response(
            context.application_understanding.add_manual_role(
                project_id,
                revision=body.revision,
                display_name=body.display_name,
            ).model_dump(mode="json"),
            status_code=201,
        )

    @router.put(
        "/api/projects/{project_id}/actions/{candidate_id}",
        response_model=ApiResponse,
    )
    def decide_action(
        project_id: str,
        candidate_id: str,
        body: CandidateDecisionRequest,
    ):
        return data_response(
            context.application_understanding.decide_action(
                project_id,
                candidate_id,
                revision=body.revision,
                decision=CandidateDecision(body.decision),
                display_name=body.display_name,
            ).model_dump(mode="json")
        )

    @router.post(
        "/api/projects/{project_id}/actions",
        response_model=ApiResponse,
    )
    def add_manual_action(project_id: str, body: ManualActionRequest):
        return data_response(
            context.application_understanding.add_manual_action(
                project_id,
                revision=body.revision,
                display_name=body.display_name,
                risk_hint=ActionRiskHint(body.risk_hint),
            ).model_dump(mode="json"),
            status_code=201,
        )

    return router


class ApplicationConnectRequest(ApiModel):
    schema_version: Literal["1"]
    source_root: str = Field(min_length=1, max_length=32_768)
    project_name: str | None = Field(default=None, min_length=1, max_length=128)


class EndpointConfirmationRequest(ApiModel):
    schema_version: Literal["1"]
    endpoint: str = Field(min_length=1, max_length=2048)
    revision: int = Field(ge=0, le=1_000_000)


class SourceAnalysisAuthorizationRequest(ApiModel):
    schema_version: Literal["1"]
    authorized: Literal[True]
    revision: int = Field(ge=0, le=1_000_000)


class SourceAnalysisRequest(ApiModel):
    schema_version: Literal["1"]
    revision: int = Field(ge=0, le=1_000_000)


class CandidateDecisionRequest(ApiModel):
    schema_version: Literal["1"]
    decision: Literal["PROPOSED", "CONFIRMED", "REJECTED"]
    display_name: str | None = Field(default=None, min_length=1, max_length=256)
    revision: int = Field(ge=0, le=1_000_000)


class ManualRoleRequest(ApiModel):
    schema_version: Literal["1"]
    display_name: str = Field(min_length=1, max_length=128)
    revision: int = Field(ge=0, le=1_000_000)


class ManualActionRequest(ApiModel):
    schema_version: Literal["1"]
    display_name: str = Field(min_length=1, max_length=256)
    risk_hint: Literal["READ", "WRITE", "DELETE", "ADMIN", "UNKNOWN"] = "UNKNOWN"
    revision: int = Field(ge=0, le=1_000_000)
